src/lib/battery/battery.h

```diff
@@ -142,10 +142,58 @@ class Battery : public ModuleParams
 	} _params{};
 
 	const int _index;
+	const uint8_t _source;
 
 	bool _first_parameter_update{true};
 	void updateParams() override;
 
+	/**
+	 * Self-contained simple lookup class for LiPo charge at a given OC voltage.
+	 * The SoC is then calculated by scaling to fit the operation max and min voltage (derived from factors such as voltage required to maintain sufficient thrust overhead).
+	 * It's important to understand these two steps and to know that this lookup table does not directly translate to the drone SoC.
+	 *
+	 * NB
+	 *   - currently const (not reading from params) just to get going quickly
+	 *   - Please ensure _lookup_size is correctly set otherwise will seg_fault
+	 *   - Please ensure _lookup is monotonic decreasing in voltage and SOC
+	 *   - We couldn't find a nuttx table which has a .size() method!  Suggestions welcome
+	 */
+	class SeesSOC
+	{
+
+	public:
+		SeesSOC(const decltype(_params) &params) : _soc_params(params) {}
+
+		// return soc [0, 1] given open circuit cell voltage
+		float ChemistryLookup(float voltage);
+		float GetSOC(float chemistry_soc);
+
+	private:
+		const decltype(_params) &_soc_params;
+		struct Lookup {
+			float _oc_voltage{0.f};
+			float _soc{0.f};
+		};
+		/// size of lookup table - NB MUST be correct or risk of seg_fault / bad lookup
+		const int _lookup_size = 12;
+		/// lookup table - NB MUST be monontonic descending
+		const Lookup _lookup[12] {
+			// Tattu 25Ah High Voltage packs; nominal 4.35V max
+			{4.350f, 1.0f},
+			{4.203f, 0.88f},
```

Comment:
May be worth 3 sig fig on the Chemistry capacity here.  Small effect, but non-zero (as there's approx 0.10 between table points, you could be up to 1% out.